Add support for building snaps

superm1 opened the pull request:
When this prefix is set fu-config, fu-engine and fu-quirks will
look for all configured content at subdirectories of this one.

This should allow usage with snaps and flatpaks by wrappers setting
this prefix at application launch time for fwupdtool.


I haven't tested this specifically in a snap or flatpak yet, but instead purged fwupd from the system and just tested with doing a deb build, switching to `build/obj-x86_64-linux-gnu` and then running fwupdtool like this:
```
sudo FWUPD_RUNTIME_PREFIX=~/src/fwupd/build/debian/tmp ./src/fwupdtool get-devices -v
```

hughsie commented:
So this is a like a runtime version of `DESTDIR`?

superm1 commented:
> So this is a like a runtime version of DESTDIR?

Yeah.  The only thing i'm not sure about is if it should be the same env variable for plugin load path as it is data.  Reason being that data might be userspecific and plugin load path probably isn't.  So maybe with plugin load path it needs to be a relative path to "launch" directory for fwupdtool.

hughsie commented:
I'm kinda erring on just supporting something like:

    DESTDIR=~/work ninja install
    DESTDIR=~/work ./src/fwupdtool install foo.cab

So it's not an _additional_ search path, it's just adding a runtime destdir to all the existing system paths.

superm1 commented:
I think the problem with that is that you don't know the runtime destdir at build time (at least with snaps).  So this is what they recommend to do: https://docs.snapcraft.io/snaps/philosophy

superm1 commented:
So according to https://docs.snapcraft.io/snaps/philosophy you won't know the runtime DESTDIR at build time.  So I don't think that approach works.

superm1 commented:
CC @ timchen119 Maybe you can comment here or get some other snap people to comment.

superm1 commented:
CC @timchen119 Maybe you can comment here or get some other snap people to comment.

hughsie commented:
I don't think "relative to the executable" path is going to work reliably, and isn't going to work when running fwupdtool in the build-dir.

Going back a bit; what are our aims here? "lets create a snap" is pointless if there's an easier way to test using a docker container for instance. A few of my aims are:

* I have RHEL 7.4, but need a newer fwupd to apply this security update to my Logitech mouse

* I'd like to test if the new fwupd release fixes this specific bug I have, without trashing my system with development-quality code

superm1 commented:
> I don't think "relative to the executable" path is going to work reliably, and isn't going to work when running fwupdtool in the build-dir.

At least in this PR I intentionally didn't add the paths for build-dir on `fwupdtool` because I am starting to feel like they're not important if we're able to create images for people to use in flatpak/snap/docker/etc.  If you want to test `fwupdtool` without an installation in place or an older one, build or grab one of those images, install it and use `fwupdtool` from it.

> Going back a bit; what are our aims here? "lets create a snap" is pointless if there's an easier way to test using a docker container for instance. A few of my aims are:

The one aim I have that isn't listed by yours is

- I have a new device only supported by a new plugin in a newer fwupd but my distro is an ancient distro and i'd like to be able to install updates to it. 

So I think the right answer for all 3 of our aims is to provide both snaps and flatpaks.  Providing docker images wouldn't work super well unless you bind mount tons of stuff in.

Does flatpak not require any of these relative directory things?  If not, then let me rework them some to directly use the snap environment variables (instead of setting them in wrapper to FWUPD_RUNTIME_PREFIX) to see how that looks.

hughsie commented on src/fu-common.c:
```diff
@@ -364,6 +364,28 @@ fu_common_add_argv (GPtrArray *argv, const gchar *fmt, ...)
+gchar*
+fu_common_get_localstatedir (void)
+{
+	const gchar *runtime_prefix = NULL;
```
needs to be "inside" the #ifdef

superm1 replied:
fixed (looks CI wasn't happy about this too)

hughsie commented on src/fu-common.c:
```diff
@@ -364,6 +364,28 @@ fu_common_add_argv (GPtrArray *argv, const gchar *fmt, ...)
+ *
+ * Gets the configured local state directory (which may be dynamic)
+ */
+gchar*
```
Space before the *

superm1 replied:
fixed

hughsie commented on src/fu-config.c:
```diff
@@ -64,6 +68,17 @@ fu_config_get_config_paths (void)
+	/* Add something with a runtime prefix */
+	runtime_prefix = g_getenv ("SNAP_USER_DATA");
+	if (runtime_prefix != NULL) {
+		runtimedir = g_build_filename (runtime_prefix, FWUPDCONFIGDIR, NULL);
```
Move the autofree here?

superm1 replied:
fixed

hughsie commented on src/fu-engine.c:
```diff
@@ -2860,16 +2860,26 @@ fu_engine_is_plugin_name_blacklisted (FuEngine *self, const gchar *name)
+#ifdef SNAP_SUPPORT
+	runtime_prefix = g_getenv ("SNAP");
+	if (runtime_prefix != NULL) {
+		plugin_directory = g_build_filename (runtime_prefix, PLUGINDIR, NULL);
```
autofree here pls

superm1 replied:
This one specifically was placed where it was because `plugin_directory` needs to be used later on in the loop.  I was trying to avoid extra `#ifdef` down there, but if you would prefer that i'll swap it around.

hughsie replied:
ahh, no worries then.

hughsie commented on src/fu-quirks.c:
```diff
@@ -311,18 +312,35 @@ gboolean
+	/* Add something with a runtime prefix */
+	runtime_prefix = g_getenv ("SNAP");
+	if (runtime_prefix != NULL) {
+		runtimedir = g_build_filename (runtime_prefix, FWUPDDATADIR, NULL);
```
same

superm1 replied:
fixed

superm1 commented on contrib/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,110 @@
+      - libgudev-1.0-0
+      - libjson-glib-1.0-0
+      - liblzma5
+    prime:
```
Regarding these package lists, I didn't put them into our package dependency system because I'm not yet sure where this makes most sense in terms of CI.  I wanted to see what you thought.
In the [documentation](https://docs.snapcraft.io/build-snaps/ci-integration) offers stuff related to both Launchpad, CircleCI and Travis CI.

Ideally this should be published to the snap store edge channel and then people can easily install it and we can have latest master always published to edge channel.

hughsie replied:
Not my area of expertise :) ¯\_(ツ)_/¯

superm1 replied:
OK :) I hope that the @lyoncore guys come by and can advise.

timchen119 replied:
@superm1 Did you try build.snapcraft.io ? Launchpad is also offering snap automatic building and uploading to store if code is on Launchpad. 

BTW I just setup https://launchpad.net/~timchen119/+snap/fwupd with this branch and found the snapcraft.yaml is not in the snap/ directory, that makes the build failed, is it ok for you to pull it out from contrib/ and make snap/ to top-level? More info on: https://github.com/canonical-websites/build.snapcraft.io/issues/784

hughsie replied:
I don't really want snap a top level directory if I can help it.

superm1 replied:
@timchen119 Actually snapcraft.yaml is dynamic because it needs to contain version information.  Is it possible to call a script before launchpad does that part of the build?  Like how debs could have launchpad packaging recipes?

I didn't use build.snapcraft.io, I just did a local build (calling `contrib/ci/snap.sh`) which prepared build directory and built snap in it.

superm1 replied:
I guess the "alternative" is to have the snapcraft.yaml in a different tree or branch with it placed in the right location, but I'm unclear how you fill in the version information from git in that scenario.

timchen119 replied:
@superm1 there is `version-script` field https://github.com/videolan/vlc/blob/master/extras/package/snap/snapcraft.yaml#L2 , another trick to fill version dynamically is to use `snapcraftctl set-version` in the snapcraft.yaml, an example here: https://github.com/ogra1/uboot-tools-snap/blob/master/snap/snapcraft.yaml#L63 ,

superm1 replied:
Thanks, those both look quite workable.

I'll do some experiment with build.snapcraft.io then.

superm1 replied:
Btw it looks that vlc does manage to keep snapcraft.yaml elsewhere in 🌲. How do they get it to build then?

superm1 commented:
@hughsie So it seems all the build services _do_ require you to have snapcraft.yaml top level.  If you would rather keep it out of tree I think that's fine.  I've adjusted the PR for that, so please take a look again.

I moved it over here for now https://github.com/superm1/fwupd-snapcraft until I can get it working with a build service and then we can figure out the right CI to do it with.  Unfortunately it's got to build on 16.04 core so it's going to be a bit more complicated to get right (Builds just fine*) on 18.04 outside the container.

superm1 commented:
I've reworked this to not add a new option but only activate when the daemon is turned off.

I've got a snap building now on 16.04 using this https://github.com/superm1/fwupd-snapcraft, but I haven't really tested it can install on devices.  Once I can prove it works properly I'll figure out it's right home.

So @hughsie can you look at the patch in this branch now? I think it should be good (and i'll squash it if you think so).

hughsie commented:
Would you mind if I respun this in a slightly different way? I've got an idea how we can make things simpler, rather than more complex, by adding more things like `fu_common_get_configdir()` which means we can move the FWUPD-prefixing out of meson and into the C, and unconditionally support both runtime DESTDIR and the various SNAP env-vars.

superm1 commented:
> Would you mind if I respun this in a slightly different way? I've got an idea how we can make things simpler, rather than more complex, by adding more things like fu_common_get_configdir() which means we can move the FWUPD-prefixing out of meson and into the C, and unconditionally support both runtime DESTDIR and the various SNAP env-vars.

Sure! That sounds fine by me.

hughsie commented:
@superm1 I was thinking something like wip/hughsie/snap-crackle-pop -- it makes things simpler in my head anyway. I have to sleep now.

superm1 commented:
OK let's close mine and discuss on yours then.
